### DISCUSSION TITLE
Do expressions support ternary operators to change their returned value?

### DISCUSSION THREAD

**Discussion by atrauzzi** [community/community#26738] [Actions]
<p>Something along these lines…</p>
<pre><code>- name: Create Release
id: create-release
uses: actions/create-release@v1
env:  
GITHUB\_TOKEN: ${{ secrets.GITHUB_TOKEN }}
with:  
tag\_name: ${{ env.VERSION }}
release\_name: ${{ env.VERSION }}
draft: false
prerelease: "${{ github.ref != 'refs/heads/master' ? 'true' : 'false' }}"
</code></pre>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>Hi <a class="mention" href="/atrauzzi">@atrauzzi</a> , </p>
<p>Expressions don’t support ternary operators. You could create a feedback for this feature in the <a href="https://support.github.com/contact/feedback?contact%5Bcategory%5D=actions" rel="nofollow noopener">Feedback form for GitHub Actions</a> .</p>
<p>As a workaround , you could add another step to set prerelease as true or false before Create Release step. In my example , I use <a href="https://github.com/actions/toolkit/blob/master/docs/commands.md#set-outputs" rel="nofollow noopener"><strong>set-output command</strong></a> to set an output variable in the front step, then use it in Create Release step .</p>
<pre><code>- name: prerelease or not
  id: pre_or_not
  run: |
    if [$REF == 'refs/heads/master']
    then
        echo "::set-output name=prerelease::true"
    else
        echo "::set-output name=prerelease::false"
    fi
  env:
    REF: ${{ github.ref }}

- name: Create Release
  id: create-release
  uses: actions/create-release@v1
  env:
    GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
  with:
    tag_name: ${{ env.VERSION }}
    release_name: ${{ env.VERSION }}
    draft: false
    prerelease: ${{ steps.pre_or_not.outputs.prerelease }}
</code></pre>

**Reply by onionhammer** [reply]
How would you set an environment url with a previous step?

**Reply by akostadinov** [reply]
FYI `::set-output` is deprecated in favor of

https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#environment-files

**Comment by Simran-B** [comment]
<p>Is a ternary operator or substitute even necessary in this case?</p>
<aside class="quote no-group" data-username="atrauzzi" data-post="1" data-topic="18114">
<div class="title">
<div class="quote-controls"></div>
 atrauzzi:</div>
<blockquote>
<p><code>${{ github.ref != 'refs/heads/master' ? 'true' : 'false' }}</code></p>
</blockquote>
</aside>
<p><code>github.ref != 'refs/heads/master'</code> evaluates to a boolean (<code>true</code> or <code>false</code>). Can’t you just use that?</p>
<pre><code class="lang-auto">prerelease: ${{ github.ref != 'refs/heads/master' }}
</code></pre>
<p>In other cases, the data type might not be right for you or undesired coercion might occur. In conjunction with <a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#operators" rel="noopener nofollow ugc">comparison operators</a>, GitHub will cast them to numbers (<code>0</code> or <code>1</code>), but in case of <a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#functions" rel="noopener nofollow ugc">functions</a>, at least some cast to string (<code>'true'</code> or <code>'false'</code>).</p>
<p>You can cast booleans to strings explicitly should it be necessary:</p>
<pre><code class="lang-auto">${{ format(github.ref != 'refs/heads/master') }}
</code></pre>
<p>I verified that the cast works like so:</p>
<pre><code class="lang-auto">- name: Cast to string
  run: |
    cat &lt;&lt;EOF
      ${{ toJSON(true) }}
      ${{ toJSON(format(true)) }}
    EOF
</code></pre>
<p>Because JSON has native boolean types, <code>true</code> remains <code>true</code> when calling <code>toJSON()</code>. The <code>format()</code> function casts it to a string, which then leads to a JSON serialization of <code>"true"</code> (note the double quotes). The wrapping <code>cat &lt;&lt;EOF ... EOF</code> preserves the quote marks, whereas <code>echo</code> would swallow them.</p>
<p>Finally, I want to point out that there is a <a href="http://rickluna.com/wp/tag/fake-ternary/" rel="noopener nofollow ugc">fake ternary</a> technique which works on GitHub:</p>
<pre><code class="lang-auto">${{ github.ref != 'refs/heads/master' &amp;&amp; 'true' || 'false' }}
</code></pre>
<p>The only limitation is that the second AND condition (here: <code>'true'</code>) must be a truthy value. If it’s falsey (<code>null</code>, <code>false</code>, <code>0</code>, <code>''</code>) then the value to the right of the OR operator will be returned, regardless of the first condition. There is no nullish coalescing operator <code>??</code> to lift this limitation.</p>
<p>Other techniques involving arrays or objects aren’t possible because expressions in Actions do not support array and object literals (yet).</p>

**Reply by nemanjaglumac** [reply]
Learning about a fake ternary here made it possible to implement custom conditional concurrency. Thank you

**Comment by Simran-B** [comment]
<p>I stand corrected: It is possible to use array/object based techniques via <code>fromJSON()</code> as an alternative to unsupported array/object literals.</p>
<pre><code class="lang-yaml">${{ fromJSON('["no", "yes"]')[github.ref != 'refs/heads/master'] }}
</code></pre>
<p>The condition evaluates to a boolean, which is implicitly cast to a number and used as array index, where <code>false</code> = <code>0</code> and <code>true</code> = <code>1</code>.</p>

**Reply by kbolino** [reply]
An adaptation of this approach is the only thing that worked well for me for handling booleans that could be unset:

```
${{ fromJSON('{"":true,"false":false,"true":true}')[format(inputs.some_bool)] }}
```

**Comment by amithm7** [comment]
<p>The fake ternary operator doesn’t work for secrets:</p>

```yaml
# works
value: ${{ github.ref != 'refs/heads/master' && 'true' || 'false' }}

# doesn't work
value: ${{ github.ref != 'refs/heads/master' && secrets.ValueA || secrets.ValueB }}
```

**Reply by samuliasmala** [reply]
This did work for me so maybe things have changed

```yml
${{ github.ref_name == 'main' && secrets.AZURE_PROD_PUBLISH_PROFILE || secrets.AZURE_DEV_PUBLISH_PROFILE }}
```

**Comment by Simran-B** [comment]
<p>Interesting. Are there secrets simple strings or like JSON/YAML objects?</p>
<p>If you want to use different secrets, then using GitHub environments might be what you want.</p>
